server/gfmd/auth_uid.c

```diff
@@ -313,6 +344,46 @@ auth_uid_to_global_username_sharedsecret(void *closure,
 	return (GFARM_ERR_NO_ERROR);
 }
 
+#ifdef HAVE_CYRUS_SASL
+
+static gfarm_error_t
+auth_uid_to_global_username_sasl(void *closure,
+	const char *auth_user_id, enum gfarm_auth_id_role *auth_user_id_rolep,
+	char **global_usernamep)
+{
+	enum gfarm_auth_id_role auth_user_id_role = *auth_user_id_rolep;
+	char *global_username;
+	struct user *u;
+	const char diag[] = "auth_uid_to_global_username_sasl";
+
+	if (auth_user_id_role != GFARM_AUTH_ID_ROLE_USER)
+		return (GFARM_ERR_AUTHENTICATION);
+
+	giant_lock();
+	u = user_tenant_lookup(auth_user_id);
+	giant_unlock();
+
+	if (u == NULL) {
+		/*
+		 * do not return GFARM_ERR_NO_SUCH_USER
+		 * to prevent information leak
+		 */
+		gflog_info(GFARM_MSG_UNFIXED,
+		    "unknown user id <%s>", auth_user_id);
+		return (GFARM_ERR_AUTHENTICATION);
+	}
+	if (global_usernamep == NULL)
+		return (GFARM_ERR_NO_ERROR);
+	global_username = strdup_log(
+		     user_auth_id(u, GFARM_AUTH_USER_ID_TYPE_SASL), diag);
+	if (global_username == NULL)
```

Comment:
this user_auth_id() function will become unnecessary, when the convertion direction is fixed.